[c10d/nccl-pg] allow user to pass process group description

shengbao-zheng opened the pull request:
Summary:
We need a way to allow user set a customized description for a process group, e.g. FSDP, PP.

Here are several use cases of user specified group_desc:
- Logging: we can easily match a log line and understand what's this collective/pg is used to.
- Pytorch traces (e.g. Kineto, Execution Trace) can benefit from the PG desc since trace analysis, benchmarks will be able to easily differentiate PG purpose like FSDP, PP.
- Lower layer collectives(e.g. NCCL) debug: we will be able to expose PG desc to NCCL communicator so NCCL layer operations can be easily correlated to a PG.

Solution: Add a group_desc field to c10d
- `group_name` remains as the unique identifier, e.g. “0”, "1"
- `group_desc` will be the user specified name.
  -  Default PG will have the desc of “default_pg”
  - If user specified the name, we set desc to it
  - If use didn’t provide, we set to “unknown” or maybe “undefined”


Differential Revision: D55781850




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

yifuwang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -434,6 +434,7 @@ def __init__(self, op: Callable, tensor: torch.Tensor, dst_tensor: Optional[torc
 # Use them through the _world object to make sure the _world override mechanism
 _pg_map: Dict[ProcessGroup, Tuple[str, Store]] = {}
 _pg_names: Dict[ProcessGroup, str] = {}
+_pg_descs: Dict[ProcessGroup, str] = {}
```
Hmm do we need this? Since we can get the descriptive name via `pg.pg_descs`, is there a reason that we want to get it via  `_pg_descs[pg]`?

IIUC these dicts here are for adding python-only properties to pgs without having to add them in C++ (I think `_pg_names` here is not needed anymore).

wconstab replied:
i dont know the history but if we can just make `_get_process_group_desc` return `pg.pg_descs` it seems better to not add all the extra python code.

shengbao-zheng replied:
Thanks for pointing this out. Yea it shouldn't be necessary to have a separate dict for pg_descs. For pg_names, I think the main purpose to help check if the group_name is unique.

shengbao-zheng replied:
addressed comments

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1446,6 +1457,8 @@ def _new_process_group_helper(
             _, prefix_store = _world.pg_map[existing_group]
             return existing_group, prefix_store
 
+    group_desc = "unknown" if group_desc is None else group_desc
```
i think "unknown" is the best we can do, even though i don't love it.

at least, when we create groups as part of an N-D device-mesh, we should make sure they get named according to the device-mesh dims automatically.  cc @wz337

shengbao-zheng replied:
this reminds me that we should also add support for new_subgroups and similar funcs that the subgroup should inherit a group_desc.

kwen2501 replied:
Is there a preference of using "unknown" over ""?
If pytorch internal coding (or even external usage bc I see the PR added a property API),
`if group_desc` seems easier than `if group_desc == "unknown"`.

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4173,14 +4187,16 @@ def test_short(self, timing_enabled):
         pg_config = t['pg_config']
         self.assertEqual(len(pg_config), 1)
-        self.assertEqual(len(pg_config['0']), self.world_size)
+        self.assertEqual(len(pg_config['0']), 3)
```
Is there a reason to use specific number instead of `self.world_size`? 
If there is a need, can you please comment what 3 represents so that a new comer could know how to modify this test?

shengbao-zheng replied:
3 represents that there are three fields for each PG in pg_config dict, i.e. "name", "desc" and "ranks". I will add a comment for this.

wconstab replied:
maybe just assert that 'name', 'desc', and 'ranks' are in the config? clearer that way

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -170,6 +170,17 @@ void ProcessGroup::setGroupName(const std::string& name) {
 
+const std::string& ProcessGroup::getGroupDesc() const {
+  TORCH_CHECK(!deviceTypeToBackend_.empty(), "ProcessGroup desc not set");
+  return deviceTypeToBackend_.begin()->second->getGroupDesc();
```
Hmm, are we going to the first backend to retrieve a ProcessGroup name?
Can't we make `pg_desc_` a member variable of `ProcessGroup`?

shengbao-zheng replied:
here pg_desc_ follows the same logic as pg_name_. So I think make them consistent would be better?

kwen2501 replied:
If `pg_name_` is done wrong, my recommendation is: do not follow it.
Later, someone can easily refactor `pg_name_` by following your right implementation.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -730,6 +730,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
       at::cuda::getNumGPUs() != 0,
       "ProcessGroupNCCL is only supported with GPUs, no GPUs found!");
   this->setGroupName(options_->group_name);
+  this->setGroupDesc(options_->group_desc);
```
Hmm, why are we setting the desc so low in the stack? It would seem only functional to NCCL.
Is it possible to set the desc at the python layer of `ProcessGroup`? (not `ProcessGroupNCCL`)

shengbao-zheng replied:
Here we want to know the group desc as well when creating ProcessGroupNCCL, for example, print set up, create log prefix, profiler, etc thus in processgroupNCCL we have this set logic. However, we do set the desc at python layer with pg. _set_group_desc after the pg creation so all backends will have the desc as well.

shengbao-zheng replied:
> The Backend class will know it bc ProcessGroup set its desc to each backend

But ProcessGroup set name/desc to each backend class **after** the initialization of a backend. In order to know the group_desc during the initialization of backend(ProcessGroupNCCL), we have to change ProcessGroupNCCL constructor. I think this will need more changes, especially some user code that calling `dist.ProcessGroupNCCL` not through c10d `new_group`.

kwen2501 replied:
See my other comment:
> why must we access this information in ProcessGroupNCCL's init function? For printing a log there? Not critical to me.

shengbao-zheng replied:
> Not only printing a log, but also pass to profiler, and flight recorder to know pg set up.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -832,7 +834,8 @@ ProcessGroupNCCL::ProcessGroupNCCL(
             << ", TORCH_NCCL_COORD_CHECK_MILSEC: " << coordCheckIntervalMilSec_
-            << ", PG Name: " << options_->group_name;
+            << ", PG Name: " << options_->group_name
+            << ", PG DESC: " << options_->group_desc;
```
nit:
``` 
<< ", PG id: " << options_->group_name
<< ", PG descriptor: " << options_->group_desc;
```

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1491,7 +1494,11 @@ std::string ProcessGroupNCCL::getNCCLWatchdogDebugInfo() {
 
 std::string ProcessGroupNCCL::createLogPrefix() const {
-  return c10::str("[PG ", uid_, " Rank ", rank_, "] ");
+  if (!pg_desc_.empty() && pg_desc_ != "unknown") {
```
this is where "" would come in handy.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -435,6 +435,7 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     int64_t split_color{0};
     std::vector<uint64_t> global_ranks_in_group;
     std::string group_name;
+    std::string group_desc;
```
Like other comments above, move this to ProcessGroup.hpp?

shengbao-zheng replied:
Since we only pass in group_desc in processGroupNCCL when creating it, I think leave it here would be better together with group_name? For other backends the group_name/desc will be set at python layer after pg creation.

kwen2501 replied:
Same as above, no need to follow something that's wrong.

shengbao-zheng replied:
We need a way to pass in group name/desc for a backend when creating it. Setting through ProcessGroup cannot allow us accessing these information in the backend init function.

shengbao-zheng replied:
Not only printing a log, but also pass to profiler, and flight record to know pg set up.

kwen2501 replied:
If the information is in the `Backend` class, which is the base for `ProcessGroupNCCL`, profiler & FR should be able to access it.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1296,7 +1296,7 @@ def init_process_group(
 
         default_pg, _ = _new_process_group_helper(
-            -1, -1, [], backend, None, group_name, timeout=timeout
+            -1, -1, [], backend, None, group_name, timeout=timeout, group_desc="default_pg"
```
nit: would "default" be better? because other descriptor may be like "dp", "tp", "pp", no "_pg" suffix. not a big deal though.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1322,6 +1322,7 @@ def init_process_group(
             pg_options=pg_options,
             timeout=timeout,
             device_id=device_id,
+            group_desc="default_pg"
```
We don't want to expose an argument here -- the default process group's name is not for setting. No change allowed (at least as of now).

shengbao-zheng replied:
we don't allow user to pass in a customized name in init_process_group, but this line of code is inside it and calls _new_process_group_helper to create the default P. Hence we need to specify the desc here.

kwen2501 replied:
Sorry, the github review format misled me. Sounds good then.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4041,6 +4060,8 @@ def new_subgroups(
             the construction of specific process groups. i.e. for the ``nccl``
             backend, ``is_high_priority_stream`` can be specified so that
             process group can pick up high priority cuda streams.
+        group_desc (str, optional): A string describing the group. Each subgroup will
+            amend subgroup id to its group_desc
```
what is a "subgroup id"? As a user, I don't see any reference to it.
And what's the need for attaching id to a user-provided `group_desc`?

shengbao-zheng replied:
in new_subgroups, each subgroup has an id. yea maybe just inheritting the group_desc is enough since subgroup_id is not exposed to users.

kwen2501 replied:
Sounds good.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4239,6 +4264,9 @@ def _get_group_tag(pg: ProcessGroup) -> str:
 def _get_process_group_name(pg: ProcessGroup) -> str:
     return _world.pg_names.get(pg, "None")
 
+def _get_process_group_desc(pg: ProcessGroup) -> str:
+    return pg.group_desc
+
```
nit: is this function used anywhere? Does it have some functionality as the `group_desc` property?

kwen2501 reviewed (approved):
Looks a lot cleaner to me! Thanks for the effort!

shengbao-zheng committed:
[c10d/nccl-pg] allow user to pass process group description (#123472)

Summary:

We need a way to allow user set a customized description for a process group, e.g. FSDP, PP.

Here are several use cases of user specified group_desc:
- Logging: we can easily match a log line and understand what's this collective/pg is used to.
- Pytorch traces (e.g. Kineto, Execution Trace) can benefit from the PG desc since trace analysis, benchmarks will be able to easily differentiate PG purpose like FSDP, PP.
- Lower layer collectives(e.g. NCCL) debug: we will be able to expose PG desc to NCCL communicator so NCCL layer operations can be easily correlated to a PG.

Solution: Add a group_desc field to c10d

Reviewed By: kwen2501

Differential Revision: D55781850
